### ISSUE TITLE
Question about updates, Yubikey usage, and UI contributions

### ISSUE THREAD

**Issue by ei8fdb** [zdia/gorilla#159]
Hi there,

I've been using Password Gorilla for about 2 years on OS X. Thanks for creating it!

Can you tell if there is possibility of updates coming? Is there an issue of contributions, or something else?

Also, is there any possibility of an update on Yubikey usage? I saw a message from @rich123 about not owning a Yubikey. I'd happily contribute a Yubikey to you if it helped?

One other thing, is there anyway in which a UI designer could help out? I'd like to contribute. I've got a few suggestions, but was wondering if there is anything of priority already?

Thanks!

**Comment by rich123** [comment]
>Thanks for creating it!

I don't get to take credit for creating it, that credit goes to Frank Pilhofer.  But in any case glad you like it.

> Can you tell if there is possibility of updates coming?

Eventually.  I'm presently testing on my systems a merged copy of the new-features and pre-160 branch.  When I'm happy that no big obvious blunders have reared their head I'll push the merged version to github.  That combination is intended to be the next version if everything goes well.

> Is there an issue of contributions, or something else?

Well, PWGorilla gets the occasional contribution, but that's it otherwise.  Any issues are more of the usual issues, real life intruding to consume time with other things not related to PWGorilla.

> Also, is there any possibility of an update on Yubikey usage? I saw a message from @rich123 about not owning a Yubikey. I'd happily contribute a Yubikey to you if it helped?

Thanks for the offer, but the issue is not the cost of the Yubi's, but the lack of any Tcl interface modules to talk to the key (beyond the basic "it pretends to be a keyboard, it types in a pre-stored password" mode that it offers).  The advanced modes require sending data to the key, and there's no Tcl interface module with which to do that.  So before one could be used for anything more than the "type a very long pre-stored random password/phrase" mode, a Tcl interface driver would have to be created.

> One other thing, is there anyway in which a UI designer could help out? I'd like to contribute. I've got a few suggestions, but was wondering if there is anything of priority already?

Well, contributions/help is always a good thing.  For the UI screens I've built so far I try to build them with a thought in mind of at least trying to not be too off-putting and reasonably intuitive (to the extent that any intuitiveness is possible).  If you have suggestions then discussing them is good.  If you can create Tk implementations that is even better. :)

**Comment by acidjunk** [comment]
Can I help with tests?

**Comment by rich123** [comment]
Yes, once we either find, or write, the necessary interface driver to talk to a Yubikey from Tcl.  Right now, the makers of the Yubikey do not provide Tcl interface code, so until someone else or we create that necessary Tcl interface code, it is impossible to talk to a Yubi from Tcl.

With that said, it is possible to utilize a Yubi in its "just a keyboard" mode and "static password storage" to be the element which contains the unlock password typed into PWGorilla to unlock a database.  Setup a static password on one of the Yubi slots, install a long (64 or 80 or 128 character) randomly generated password into the static slot.  Then use the Yubi to "type" that password into PWGorilla as if you had typed it yourself from the keyboard.

This mode is not the 2-factor mode that is possible with the proper interface code, but it would allow you to use a significantly longer, and totally random, password as your unlock password than anything you'd be willing to type yourself.  Which means that if someone were to steal your laptop or other device with the file, they would have an even harder time of trying to brute-force the unlock password.

As well, I grabbed the PWSafe source that has the Yubi interface.  They are using the Yubi mode I expected (SHA1-HMAC).  It works by having PWSafe install a random HMAC key into the Yubi, then passing your enterend password to the Yubi to be SHA1-HMAC scrambled using the pre-stored random HMAC key, and the returned HMAC output from the Yubi is then the unlock password for the password file.

So the 'how' to support the device is there, what we lack at the moment is the code to interface it to Tcl.
